File: problems/tier2-intermediate/018-twitter/solution.go
// Twitter — post tweets, follow/unfollow, k-way merged news feed (Go port).
package main

import (
	"sort"
	"strconv"
)

type TwitterOp struct {
	kind string
	i1   int
	i2   int
}

type tweetRec struct {
	tid int
	ts  int
}

type twitter struct {
	time    int
	tweets  map[int][]tweetRec
	follows map[int]map[int]bool
}

func newTwitter() *twitter {
	return &twitter{
		time:    0,
		tweets:  map[int][]tweetRec{},
		follows: map[int]map[int]bool{},
	}
}

func (t *twitter) postTweet(userID, tweetID int) {
	t.tweets[userID] = append(t.tweets[userID], tweetRec{tweetID, t.time})
	t.time++
}

func (t *twitter) getNewsFeed(userID int) []int {
	users := map[int]bool{userID: true}
	for f := range t.follows[userID] {
		users[f] = true
	}
	// Collect the candidate tweets. Timestamps are globally unique (time is a
	// monotonically increasing counter), so a global sort by descending
	// timestamp reproduces the k-way merge ordering exactly.
	cand := []tweetRec{}
	for uid := range users {
		cand = append(cand, t.tweets[uid]...)
	}
	sort.Slice(cand, func(i, j int) bool { return cand[i].ts > cand[j].ts })
	result := []int{}
	for i := 0; i < len(cand) && len(result) < 10; i++ {
		result = append(result, cand[i].tid)
	}
	return result
}

func (t *twitter) follow(followerID, followeeID int) {
	if followerID == followeeID {
		return
	}
	if t.follows[followerID] == nil {
		t.follows[followerID] = map[int]bool{}
	}
	t.follows[followerID][followeeID] = true
}

func (t *twitter) unfollow(followerID, followeeID int) {
	if s := t.follows[followerID]; s != nil {
		delete(s, followeeID)
	}
}

func twitter_simulate(ops []TwitterOp) []string {
	out := []string{}
	tw := newTwitter()
	for _, op := range ops {
		switch op.kind {
		case "new":
			tw = newTwitter()
			out = append(out, "ok")
		case "post":
			tw.postTweet(op.i1, op.i2)
			out = append(out, "ok")
		case "follow":
			tw.follow(op.i1, op.i2)
			out = append(out, "ok")
		case "unfollow":
			tw.unfollow(op.i1, op.i2)
			out = append(out, "ok")
		case "feed_size":
			out = append(out, strconv.Itoa(len(tw.getNewsFeed(op.i1))))
		case "feed_at":
			f := tw.getNewsFeed(op.i1)
			if op.i2 >= 0 && op.i2 < len(f) {
				out = append(out, strconv.Itoa(f[op.i2]))
			} else {
				out = append(out, "-1")
			}
		default:
			out = append(out, "unknown:"+op.kind)
		}
	}
	return out
}

File: problems/tier2-intermediate/018-twitter/solution.java
// Twitter — Solution (Java)
import java.util.*;

class TwitterOp {
    public String kind;
    public int i1, i2;
    public TwitterOp(String kind, int i1, int i2) { this.kind = kind; this.i1 = i1; this.i2 = i2; }
}

class Tweet {
    public int tweetId;
    public int timestamp;
    public Tweet(int t, int ts) { tweetId = t; timestamp = ts; }
}

class Twitter {
    int time = 0;
    Map<Integer, List<Tweet>> tweets = new HashMap<>();
    Map<Integer, Set<Integer>> follows = new HashMap<>();

    public void postTweet(int userId, int tweetId) {
        tweets.computeIfAbsent(userId, k -> new ArrayList<>()).add(new Tweet(tweetId, time++));
    }

    public List<Integer> getNewsFeed(int userId) {
        Set<Integer> users = new HashSet<>();
        users.add(userId);
        Set<Integer> f = follows.get(userId);
        if (f != null) users.addAll(f);

        // Heap entries: int[]{timestamp, tweetId, userId, index}
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(b[0], a[0]));
        for (int uid : users) {
            List<Tweet> ts = tweets.get(uid);
            if (ts != null && !ts.isEmpty()) {
                int idx = ts.size() - 1;
                pq.offer(new int[]{ts.get(idx).timestamp, ts.get(idx).tweetId, uid, idx});
            }
        }

        List<Integer> result = new ArrayList<>();
        while (!pq.isEmpty() && result.size() < 10) {
            int[] top = pq.poll();
            int idx = top[3];
            int uid = top[2];
            result.add(top[1]);
            if (idx > 0) {
                int next = idx - 1;
                Tweet nt = tweets.get(uid).get(next);
                pq.offer(new int[]{nt.timestamp, nt.tweetId, uid, next});
            }
        }
        return result;
    }

    public void follow(int followerId, int followeeId) {
        if (followerId != followeeId) {
            follows.computeIfAbsent(followerId, k -> new HashSet<>()).add(followeeId);
        }
    }

    public void unfollow(int followerId, int followeeId) {
        Set<Integer> set = follows.get(followerId);
        if (set != null) set.remove(followeeId);
    }
}

public class Solution {
    public static List<String> twitter_simulate(List<TwitterOp> ops) {
        List<String> out = new ArrayList<>();
        Twitter tw = new Twitter();
        for (TwitterOp op : ops) {
            String k = op.kind;
            if ("new".equals(k)) {
                tw = new Twitter();
                out.add("ok");
            } else if ("post".equals(k)) {
                tw.postTweet(op.i1, op.i2);
                out.add("ok");
            } else if ("follow".equals(k)) {
                tw.follow(op.i1, op.i2);
                out.add("ok");
            } else if ("unfollow".equals(k)) {
                tw.unfollow(op.i1, op.i2);
                out.add("ok");
            } else if ("feed_size".equals(k)) {
                out.add(Integer.toString(tw.getNewsFeed(op.i1).size()));
            } else if ("feed_at".equals(k)) {
                List<Integer> f = tw.getNewsFeed(op.i1);
                out.add(op.i2 >= 0 && op.i2 < f.size() ? Integer.toString(f.get(op.i2)) : "-1");
            } else {
                out.add("unknown:" + k);
            }
        }
        return out;
    }
}
